File: sdk/typescript/packages/tensorzero-sdk/test/e2e.test.ts
// Modified by Delta-AI under Apache 2.0
import { describe, expect, it } from "vitest";
import { createTensorZeroClient, TaskNotFoundError } from "../src/index.js";

const GATEWAY = process.env["TZ_E2E_GATEWAY"];
const API_KEY = process.env["TZ_E2E_KEY"];
const RUN_E2E = Boolean(GATEWAY && API_KEY);
const MODEL = "deepseek-v4-flash";

const client = RUN_E2E
  ? createTensorZeroClient({ baseURL: GATEWAY!, apiKey: API_KEY! })
  : undefined;

describe.skipIf(!RUN_E2E)("e2e against a real gateway", () => {
  it("status() and health() are reachable without auth failures", async () => {
    // /status is the liveness endpoint and returns 200 when the gateway is up.
    const status = (await client!.status()) as { status?: string };
    expect(status.status).toBe("ok");
    // /health reports per-component health and returns 503 when any component
    // (e.g. valkey) is degraded — both outcomes are valid wire behavior.
    try {
      await client!.health();
    } catch (error) {
      expect(error).toMatchObject({ name: "TensorZeroHttpError", statusCode: 503 });
      expect((error as { body?: { gateway?: string } }).body?.gateway).toBe("ok");
    }
  });

  it("submit → waitForCompletion → completed chat completion", async () => {
    const { taskId } = await client!.submitChatCompletion({
      model: MODEL,
      messages: [{ role: "user", content: "Say hello in one word." }],
      max_tokens: 32,
    });
    expect(typeof taskId).toBe("string");

    const final = await client!.waitForCompletion(taskId, {
      intervalMs: 500,
      maxIntervalMs: 2_000,
      timeoutMs: 120_000,
    });
    expect(final.status).toBe("completed");
    if (final.status !== "completed") return;

    const response = final.response as {
      object?: string;
      choices?: Array<{ message?: { content?: string } }>;
    };
    expect(response.object).toBe("chat.completion");
    const content = response.choices?.[0]?.message?.content;
    expect(typeof content).toBe("string");
    expect(content!.length).toBeGreaterThan(0);
  }, 150_000);

  it("getTask reports legal intermediate states", async () => {
    const { taskId } = await client!.submitChatCompletion({
      model: MODEL,
      messages: [{ role: "user", content: "Count to ten slowly." }],
      max_tokens: 64,
    });

    // Poll directly and collect every state observed along the way.
    const seen: string[] = [];
    for (;;) {
      const status = await client!.getTask(taskId);
      expect(status.taskId).toBe(taskId);
      seen.push(status.status);
      if (status.status === "completed") break;
      if (status.status === "failed" || status.status === "cancelled") {
        throw new Error(`task unexpectedly ${status.status}`);
      }
      await new Promise((resolve) => setTimeout(resolve, 200));
    }

    expect(seen.at(-1)).toBe("completed");
    for (const state of seen) {
      expect(["queued", "running", "completed"]).toContain(state);
    }
  }, 150_000);

  it("streamTask relays wire chunks and ends quietly", async () => {
    const { taskId } = await client!.submitChatCompletion({
      model: MODEL,
      messages: [{ role: "user", content: "Say hello in one word." }],
      max_tokens: 32,
    });

    const events = [];
    for await (const event of client!.streamTask(taskId, {
      reconnectIntervalMs: 200,
      fallbackPollIntervalMs: 200,
    })) {
      events.push(event);
    }

    // Wire-faithful: every yielded item is an "event" frame from the wire;
    // no synthetic terminal item exists in the type or at runtime.
    expect(events.length).toBeGreaterThan(0);
    let sawChunk = false;
    let sawDone = false;
    for (const event of events) {
      expect(event.type).toBe("event");
      if (event.data === "[DONE]") {
        sawDone = true;
        continue;
      }
      const parsed = event.json as { object?: string } | undefined;
      expect(parsed).toBeDefined();
      expect(parsed!.object).toBe("chat.completion.chunk");
      sawChunk = true;
    }
    expect(sawChunk).toBe(true);
    expect(sawDone).toBe(true);

    // The final result comes from getTask, not the stream.
    const status = await client!.getTask(taskId);
    expect(status.status).toBe("completed");
  }, 150_000);

  it("getTask on an unknown task id throws TaskNotFoundError", async () => {
    await expect(
      client!.getTask("0190f9c4-8e3a-7b3d-9c1e-2f4a5b6c7d8e"),
    ).rejects.toBeInstanceOf(TaskNotFoundError);
  });

  it("a bad model name reaches a failed terminal state", async () => {
    const { taskId } = await client!.submitChatCompletion({
      model: "definitely-not-a-model",
      messages: [{ role: "user", content: "hi" }],
      max_tokens: 8,
    });

    const final = await client!.waitForCompletion(taskId, {
      intervalMs: 500,
      maxIntervalMs: 2_000,
      timeoutMs: 60_000,
    });
    expect(final.status).toBe("failed");
    if (final.status !== "failed") return;
    const message = JSON.stringify(final.error);
    expect(message).toContain("not found in model table");
  }, 90_000);
});

File: sdk/go/e2e_test.go
// Modified by Delta-AI under Apache 2.0
package tensorzero

import (
	"context"
	"encoding/json"
	"errors"
	"os"
	"strings"
	"testing"
	"time"
)

// End-to-end tests against a real gateway. They run only when both
// TZ_E2E_KEY and TZ_E2E_GATEWAY are set (and are skipped under -short). The
// API key is read from the environment only — never commit it.
//
//	TZ_E2E_GATEWAY=https://gateway.example.com TZ_E2E_KEY=sk-... go test -run E2E -v
func e2eClient(t *testing.T) *Client {
	t.Helper()
	if testing.Short() {
		t.Skip("skipping e2e test in -short mode")
	}
	key := os.Getenv("TZ_E2E_KEY")
	gateway := os.Getenv("TZ_E2E_GATEWAY")
	if key == "" || gateway == "" {
		t.Skip("TZ_E2E_KEY and TZ_E2E_GATEWAY must both be set; skipping e2e test")
	}
	return NewClient(gateway, WithAPIKey(key))
}

func e2eContext(t *testing.T) context.Context {
	t.Helper()
	ctx, cancel := context.WithTimeout(context.Background(), 3*time.Minute)
	t.Cleanup(cancel)
	return ctx
}

var e2eChatBody = json.RawMessage(`{
	"model": "deepseek-v4-flash",
	"messages": [{"role": "user", "content": "Reply with the single word: pong"}],
	"max_tokens": 64
}`)

// chatCompletion mirrors the fields of a chat.completion response/chunk that
// these tests assert on.
type chatCompletion struct {
	ID      string `json:"id"`
	Object  string `json:"object"`
	Choices []struct {
		Message struct {
			Role    string `json:"role"`
			Content string `json:"content"`
		} `json:"message"`
		Delta struct {
			Role    string `json:"role"`
			Content string `json:"content"`
		} `json:"delta"`
	} `json:"choices"`
}

func TestE2EStatusAndHealth(t *testing.T) {
	client := e2eClient(t)
	ctx := e2eContext(t)

	status, err := client.Status(ctx)
	if err != nil {
		t.Fatalf("Status: %v", err)
	}
	if !json.Valid(status) || !strings.Contains(string(status), "ok") {
		t.Fatalf("status body = %s", status)
	}
	health, err := client.Health(ctx)
	if err != nil {
		t.Fatalf("Health: %v", err)
	}
	t.Logf("status=%s health=%s", status, health)
}

func TestE2ESubmitWaitCompleted(t *testing.T) {
	client := e2eClient(t)
	ctx := e2eContext(t)

	taskID, err := client.SubmitChatCompletion(ctx, e2eChatBody)
	if err != nil {
		t.Fatalf("submit: %v", err)
	}
	t.Logf("task_id=%s", taskID)

	// The first read may catch an intermediate state; it must be a valid one.
	first, err := client.GetTask(ctx, taskID)
	if err != nil {
		t.Fatalf("GetTask: %v", err)
	}
	switch first.Status {
	case TaskQueued, TaskRunning, TaskCompleted:
		t.Logf("intermediate status=%s", first.Status)
	default:
		t.Fatalf("unexpected intermediate status %q", first.Status)
	}

	final, err := client.WaitForCompletion(ctx, taskID,
		WithInitialInterval(500*time.Millisecond),
		WithMaxInterval(3*time.Second),
		WithTimeout(2*time.Minute),
	)
	if err != nil {
		t.Fatalf("WaitForCompletion: %v", err)
	}
	if final.Status != TaskCompleted {
		t.Fatalf("status = %s, error = %s", final.Status, final.Error)
	}

	var resp chatCompletion
	if err := json.Unmarshal(final.Response, &resp); err != nil {
		t.Fatalf("response is not a chat.completion: %v\n%s", err, final.Response)
	}
	if resp.Object != "chat.completion" {
		t.Fatalf("object = %q, want chat.completion", resp.Object)
	}
	if len(resp.Choices) == 0 || strings.TrimSpace(resp.Choices[0].Message.Content) == "" {
		t.Fatalf("empty choices/message content in %s", final.Response)
	}
	t.Logf("content=%q", resp.Choices[0].Message.Content)
}

func TestE2EStreamTaskWireFaithful(t *testing.T) {
	client := e2eClient(t)
	ctx := e2eContext(t)

	// A longer-running completion leaves room for the stream to follow live
	// even across transient gateway retries of the stream endpoint.
	body := json.RawMessage(`{
		"model": "deepseek-v4-flash",
		"messages": [{"role": "user", "content": "Count from 1 to 30, one number per line."}],
		"max_tokens": 400
	}`)
	taskID, err := client.SubmitChatCompletion(ctx, body)
	if err != nil {
		t.Fatalf("submit: %v", err)
	}

	events, errs := client.StreamTask(ctx, taskID)
	var chunks, doneSentinels int
	for ev := range events {
		// Wire-faithful: the gateway never emits events with these names;
		// any of them would mean the SDK synthesized a terminal event.
		switch ev.Event {
		case "completed", "failed", "cancelled":
			t.Fatalf("synthetic terminal event leaked into stream: %+v", ev)
		}
		// The gateway forwards OpenAI's `data: [DONE]` sentinel verbatim.
		if string(ev.Data) == "[DONE]" {
			doneSentinels++
			continue
		}
		var chunk chatCompletion
		if err := json.Unmarshal(ev.Data, &chunk); err != nil {
			t.Fatalf("event %q data is not a chat.completion.chunk: %v\n%s", ev.Event, err, ev.Data)
		}
		if chunk.Object != "chat.completion.chunk" {
			t.Fatalf("object = %q, want chat.completion.chunk", chunk.Object)
		}
		chunks++
	}
	if err := <-errs; err != nil {
		t.Fatalf("stream error: %v", err)
	}
	if chunks == 0 {
		t.Fatal("expected at least one chunk event")
	}
	t.Logf("received %d chunk events, %d [DONE] sentinels; stream ended quietly", chunks, doneSentinels)

	// The final result comes from GetTask, not the stream.
	final, err := client.GetTask(ctx, taskID)
	if err != nil {
		t.Fatalf("GetTask: %v", err)
	}
	if final.Status != TaskCompleted {
		t.Fatalf("status = %s after stream end", final.Status)
	}
}

func TestE2ETaskNotFound(t *testing.T) {
	client := e2eClient(t)
	_, err := client.GetTask(e2eContext(t), "00000000-0000-0000-0000-000000000000")
	var notFound *TaskNotFoundError
	if !errors.As(err, &notFound) {
		t.Fatalf("err = %v, want TaskNotFoundError", err)
	}
}

func TestE2EBadModelFails(t *testing.T) {
	client := e2eClient(t)
	ctx := e2eContext(t)

	body := json.RawMessage(`{
		"model": "definitely-missing-model",
		"messages": [{"role": "user", "content": "hi"}]
	}`)
	taskID, err := client.SubmitChatCompletion(ctx, body)
	if err != nil {
		t.Fatalf("submit: %v", err)
	}
	final, err := client.WaitForCompletion(ctx, taskID, WithTimeout(2*time.Minute))
	if err != nil {
		t.Fatalf("WaitForCompletion: %v", err)
	}
	if final.Status != TaskFailed {
		t.Fatalf("status = %s, want failed", final.Status)
	}
	if !strings.Contains(string(final.Error), "not found in model table") {
		t.Fatalf("error = %s, want it to mention the model table", final.Error)
	}
}
